### ISSUE TITLE
Version 0.46 working directory problem

### ISSUE THREAD

**Issue by yanne** [robotframework/RIDE#1039]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=1000) by bocadillodeatun on 2 Jul 2012*

With version 0.46, I cannot run my tests anymore.

I installed "ride.py" with pip, so the actual binary can be found in 

  /usr/local/bin/ride.py

...which has been added to my $PATH.

My working dir is structured like this:

  /data/work
    tests
    resources
    input_data

So, I usually do this to run my tests:

  cd /data/work
  ride.py tests
  [Run]

However, version 0.46 fails with this message:

  working directory: /data/work
  command: pybot --variablefile input_data/BENCH040.py /data/work/tests

  [ ERROR ] Importing variable file 'input_data/BENCH040.py' failed: File or directory does not exist.

It looks like, even if ride says that "/data/work" is the current directory, the _real_ one is "/data/work/tests" and I need to change the "--variablefile" argument so that it reads "../input_data/BENCH040.py" now.

Shouldn't paths be relative to the place where I start "ride" from and _not_ the place where tests are located?

Even worse, with this change, I would need to change _all_ my "Import" statements because  they expect "pybot" to be executed in the parent folder that hosts the "tests" subfolder.

In other words: "ride" should execute "pybot" from the same folder where "ride" is invoked" and not from where the "tests" are located.

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=1000#c1) by @jussimalinen on 2 Jul 2012*


This was specifically changed so that the working directory during the run is where your tests root is. (See <a class="closed_ref" href="https://code.google.com/p/robotframework-ride/issues/detail?id=974" title="When running test cases, the working directory is where RIDE is started, not where the test cases are located"> issue 974 </a>)

If you want your tests to work as expected, you should start your ride:

cd /data/work
ride.py .

Does this work for you?

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=1000#c2) by @jussimalinen on 2 Jul 2012*


I am at loss what the working directory printout does at the beginning of the output.

It is the current working directory in the OS, but it almost never is the working directory for the test run (which starts in a new process and has usually a completely different working directory.)

Seems rather confusing. I will remove the printout because it seems more like a debug printout that has been left accidentally.

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=1000#c4) by bocadillodeatun on 2 Jul 2012*


So, right now, "pybot" is always executed from the "tests" folder, am I right?

Are these two commands equivalent?

  1) ride-0.46.py tests_folder

  2) cd tests_folder; ride-0.45.py .


If so, the problem with version 0.46 is that tests must be "prepared" to be run from the root tests folder (ie. all the "Import" statements must be relative to the tests root folder).

For those of us who were executing "pybot" from a parent folder that contained both the "tests" and "resources" subfolders, we cannot use ride.py version 0.46 unless we change all the Import commands to include a "../" prefix.

This is probably better explained with an example :)

With version 0.45, from the parent folder, I can do this:

  # ride.py tests

...which would execute "pybot" from the same folder I'm calling "ride.py".

All my test cases "Import" files from "resources/", and it works.

However, because version 0.46 executes "pybot" from inside "tests", my test cases cannot find the "Import"'ed files unless I change their path to "../resources/" on each of them.

I'm not saying I cannot do this (after all, it's a quick find&amp;replace :), however, it feels... I don't know... awkward.

I mean, applications tend to "preserve" their working directory when executed. In the case of robot it also brings some benefits: you can execute several "ride.py" binaries from different folders against the same tests and have "pybot" output the different files corresponding to the different executions in each of those different folders (instead of having everything mixed up in the tests folder) (Alas, the same result could be achieved with "--output" on each execution, but you get the point :)

Regarding <a class="closed_ref" href="https://code.google.com/p/robotframework-ride/issues/detail?id=974" title="When running test cases, the working directory is where RIDE is started, not where the test cases are located"> issue 974 </a> (<a href="https://code.google.com/p/robotframework-ride/issues/detail?id=974" rel="nofollow">https://code.google.com/p/robotframework-ride/issues/detail?id=974</a>), I would have solved the root problem by fixing the $PATH variable (so that the user can execute "ride.py" from any place) and not by making ride.py changing the 
 working dir for him.

Think of this in this other way:

  1) With version 0.45 you could BOTH choose where to host your
     test cases and where to place your libraries and resources.

  2) With version 0.46 your libraries and resources MUST be linked
     to your test cases, because pybot will always be executed
     from the same place.

Does this sound reasonable? :)

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=1000#c5) by bocadillodeatun on 2 Jul 2012*


I think there might be a solution to this that would make everyone happy:

  * If you run "ride.py" without arguments (and then "Open" the tests folder from within ride), change to that folder before running "pybot" (that's, I guess, the typical "Windows user" situation)

  * However, if you run "ride.py &lt;test_folder&gt;", then assume the user knows what she is doing and do not change the working dir.

This solution is a bit "hacky"/"inconsistent", but it allows "pybot" to be run from any folder if the user requires so.

This is just an idea. I think I could also live if you leave version 0.46 as it is now :) I'll just change my "Import" statements.

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=1000#c6) by @jussimalinen on 4 Jul 2012*


The guys who implemented this feature are coming back from holidays in August. They should take a look at this, I definitely don't have time in July to work on this.

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=1000#c8) by @mkorpela on 6 Aug 2012*


You can also use the pythonpath setting in RIDE to point to a location that should be used as a place to search for libraries.

Tools \ Preferences \ Importing

Does this solve your issues with the new working directory?

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=1000#c9) by bocadillodeatun on 7 Aug 2012*


It does!

However, I still had to change some "hard" references to the old folder structure.

For example, I had a keyword that was directly accessing file "input_data/whatever.txt", and I had to change that into "../input_data/whatever.txt".

That's why, even if this works, I still cannot isolate the tests from the configuration environment. This is how I usually work:

  /data
    run_configuration_1   &lt;-- This is where I
      input_data              run "pybot" from 
    run_configuration_2   &lt;-- the command line
      input_data
    ...

  /home
    src
      test  &lt;-- This is where test cases are located


So, I "cd" into the "run_configuration" folder
I'm interested in (depending on what hardware
I what to test) and then run something like this:

  pybot --variablefile input_data/something.py /home/src/tests

The problem is that the solution from the previous post
works for "Import"'ed files, but not (obviously :) ) for 
paths being directly used by keywords.

Image, for instance, one keyword which does this:

  Cat File    input_data/data.txt

When I run pybot from the console (inside any of the
"configuration" folders), it does work, but I cannot
make it work with ride because the "pybot execution folder"
information is now lost.
From the point of view of ride, the "tests" folder is
all it knows about.

Maybe, in the Tools \ Preferences \ Importing menu a new
option to tick could be added that reads something like
"Do not change execution folder" or something like that,
which makes the "old way" the default one.

What do you think?


PS: I could also export a "${RUNTIME}" variable in the file
that is initially given in the "--variablefile" argument and
then make all paths relative to that, it's just that I think
that the other option is "cleaner" :)

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=1000#c10) by bocadillodeatun on 7 Aug 2012*


I just thought in a simpler way to put this :)

When I call pybot from the command line I have TWO pieces of information regarding folders:

   sauron@&#8288;mordor:/data/run_configuration_1$ pybot /home/src/test
                 -------------------------        --------------
                           A                             B

..and I can "use" both of them ("A" is where all paths are referenced and "B" is where tests are located)

But when using ride, "A" is lost.

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=1000#c11) by @mkorpela on 7 Aug 2012*


Another option is just to add a custom start script to the base directory and use it to start robot run (just pass all the commandline arguments to it).

Custom scripts working directory is the directory where the custom script is located.

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=1000#c12) by @mkorpela on 7 Aug 2012*


@&#8288;bocadillodeatun:

"sauron@&#8288;mordor" LOL

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=1000#c13) by bocadillodeatun on 7 Aug 2012*


It works! Thanks.

Let me explain here what I did in case anyone else ever needs to do the same:

First I create a "run.sh" script on each "run_configuration" folder which contains these two lines...

  #!/usr/bin/env bash
  pybot $@&#8288;

Then, inside ride, I change the "execution profile" (in the "run tab") from "pybot" to "custom script" and point to this script (the one in the "run_configuration" folder I'm interested in).

Finally, I run the tests. The folder where the "run.sh" script is located is preserved and everything works as expected.

Thanks!

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=1000#c14) by @kontulai on 8 Aug 2012*


Closing this as WontFix. Lets reconsider if this change causes more problems.
